### PR TITLE
Add block sparse linear and locally connected layers

### PR DESCRIPTION
**Pull request by clane9**

### CODE REVIEW

**Comment by clane9** [comment]
I updated the implementation:

- Use pytorch native blocksparse tensors following [this blog](https://pytorch.org/blog/speeding-up-vits/) rather than triton blocksparse, which seems not yet stable. 
- Represent weights in `BlockSparseLinear` as block-sparse tensor rather than dense tensor. This will save a lot of gpu memory.
- Change `BlockSparseLocallyConnected` interface to more closely match `nn.Conv2d`. Except remove support for padding and stride. For now we should be able to restrict to `padding="same"`, `stride=1` (@alismil what do you think?).
- Rewrite function for constructing local connectivity matrix to directly construct a sparse rather than dense matrix. Use vectorized operations rather than for loops for construction. This should save a lot of memory and run faster.
- Add support for multiple input and output channels and depthwise convolution. The channels axis can either be first or last. For depthwise convolution, first should be more efficient (more block sparsity).

File: columnformers/models/layers.py [hunk 1]
```diff
@@ -280,3 +283,232 @@ def init_weights(module: nn.Module):
...
+        )
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        in_pattern = "n (h w) c" if self.in_shape == "nlc" else "n c h w"
+        out_pattern = "n (h w c)" if self.channels_last else "n (c h w)"
+        output = rearrange(
+            input, f"{in_pattern} -> {out_pattern}", h=self.height, w=self.height
+        )
+        output = self.bsl(output)
+        output = rearrange(
```
Comment (alismil) [inline columnformers/models/layers.py]: Does it make sense to keep rearranging the input on each forward pass? Might be slightly more efficient if we assume all inputs are batches of flattened images and then have a separate function to rearrange when necessary (e.g. to visualise activations etc)?

**Comment by alismil** [comment]
> For now we should be able to restrict to `padding="same"`, `stride=1` (@alismil what do you think?).

If we want to use this for All TNNs, to make the dimensions work out we have to use different strides for each layer

File: columnformers/models/layers.py (hunk 1 above)
Comment (clane9) [reply columnformers/models/layers.py]: Yes good point. If you have `in_shape="nlc"` and `channels_last=True` (or `in_shape="nchw"` and `channels_last=False`) the rearrange will just produce a new view without actually moving the data. Otherwise however, it will move the data a lot like you say.

My thinking was that the rearranging helps the `BlockSparseLocallyConnected` be consistent with `nn.Conv2d` (when `in_shape="nchw"`) and with `nn.Linear` (when `in_shape="nlc"`).

What if we add a third `in_shape` option `nd` that assumes the input is already flattened?

**Comment by clane9** [comment]
> If we want to use this for All TNNs, to make the dimensions work out we have to use different strides for each layer

Ok good point. Then maybe we should add back the support you had for strides and padding to `_sparse_local_connectivity`. Hopefully it doesn't complicate things too much.

File: columnformers/models/layers.py (hunk 1 above)
Comment (alismil) [reply columnformers/models/layers.py]: Yes I think that's a good idea?

File: columnformers/models/layers.py [hunk 2]
```diff
@@ -280,3 +283,240 @@ def init_weights(module: nn.Module):
...
+    ), "in channels must match out channels for depthwise"
+    N = height * height
+
+    # ij indices of input grid
+    # (h^2, 2)
+    col_indices = torch.cartesian_prod(torch.arange(height), torch.arange(height))
+
+    # conv kernel index offsets. note that the kernel width is required to be odd.
+    # (k^2, 2)
+    kernel_half_width = (kernel_size - 1) // 2
```
Comment (alismil) [inline columnformers/models/layers.py:463]: This is a bit restrictive, can we adapt this to also include even kernel size by doing something like this?

```
if kernel_size % 2 == 0:
    kernel_half_width = kernel_size // 2
    kernel_indices = torch.cartesian_prod(
        torch.arange(-kernel_half_width, kernel_half_width),
        torch.arange(-kernel_half_width, kernel_half_width),
    )
else:
    kernel_half_width = (kernel_size - 1) // 2
    kernel_indices = torch.cartesian_prod(
        torch.arange(-kernel_half_width, kernel_half_width + 1),
        torch.arange(-kernel_half_width, kernel_half_width + 1),
    )
```

Comment (clane9) [reply columnformers/models/layers.py:463]: Yes I think you're right, something like this would probably be better. Although it feels like it should be possible to make the code shorter.

More generally, it would probably be best to have exactly the same interface and behavior as [Conv2d](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html). What I have now takes a few shortcuts.
